cpp/serve/engine_actions/new_request_prefill.cc

```diff
@@ -117,83 +120,130 @@ class NewRequestPrefillActionObj : public EngineActionObj {
+        hidden_states_for_sample =
+            models_[0]->BatchGetHidden(hidden_states, request_internal_ids, prefill_lengths);
+        logits_for_sample = models_[0]->GetLogits(hidden_states_for_sample);
       }
-
-      for (int child_idx : rsentry->child_indices) {
-        if (rstates_of_entries[i]->entries[child_idx]->mstates[0]->committed_tokens.empty()) {
-          // If rstates_of_entries[i]->entries[child_idx] has no committed token,
-          // the prefill of the current rsentry will unblock
-          // rstates_of_entries[i]->entries[child_idx],
-          // and thus we want to sample a token for rstates_of_entries[i]->entries[child_idx].
-          sample_indices.push_back(i);
-          rsentries_for_sample.push_back(rstates_of_entries[i]->entries[child_idx]);
-          request_ids.push_back(rsentry->request->id);
-          generation_cfg.push_back(rsentry->request->generation_cfg);
-          rngs.push_back(&rstates_of_entries[i]->entries[child_idx]->rng);
-
-          ICHECK(rstates_of_entries[i]->entries[child_idx]->status == RequestStateStatus::kPending);
-          rstates_of_entries[i]->entries[child_idx]->status = RequestStateStatus::kAlive;
-          for (int model_id = 0; model_id < static_cast<int>(models_.size()); ++model_id) {
+      ICHECK_EQ(hidden_states_for_sample->ndim, 3);
+      ICHECK_EQ(hidden_states_for_sample->shape[0], 1);
+      ICHECK_EQ(hidden_states_for_sample->shape[1], num_rsentries);
+
+      // - Update logits.
+      ICHECK(logits_for_sample.defined());
+      Array<GenerationConfig> generation_cfg;
+      Array<RequestModelState> mstates_for_logitproc;
+      generation_cfg.reserve(num_rsentries);
+      mstates_for_logitproc.reserve(num_rsentries);
+      for (int i = 0; i < num_rsentries; ++i) {
+        generation_cfg.push_back(prefill_inputs[i].rsentry->request->generation_cfg);
+        mstates_for_logitproc.push_back(
+            prefill_inputs[i].rsentry->mstates[sample_with_base ? 0 : model_id]);
+      }
+      logits_for_sample = logits_for_sample.CreateView({num_rsentries, logits_for_sample->shape[2]},
+                                                       logits_for_sample->dtype);
+      logit_processor_->InplaceUpdateLogits(logits_for_sample, generation_cfg,
+                                            mstates_for_logitproc, request_ids);
+
+      // - Compute probability distributions.
+      NDArray probs_on_device =
+          logit_processor_->ComputeProbsFromLogits(logits_for_sample, generation_cfg, request_ids);
+
+      // - Sample tokens.
+      //   For prefill_inputs which have children, sample
+      //   one token for each rstate that is depending.
+      //   Otherwise, sample a token for the current rstate.
+      std::vector<int> sample_indices;
+      std::vector<RequestStateEntry> rsentries_for_sample;
+      std::vector<RandomGenerator*> rngs;
+      sample_indices.reserve(num_rsentries);
+      rsentries_for_sample.reserve(num_rsentries);
```

Comment:
Why we move the sampling into the outer for-loop of models? Does this mean that we may sample different tokens for different models?

Reply:
For base model, sample and commit a token to all the models. For each small model, sample a draft token.